rust/datafusion/src/physical_plan/filter.rs

```diff
@@ -103,25 +103,23 @@ impl ExecutionPlan for FilterExec {
     }
 
     async fn execute(&self, partition: usize) -> Result<SendableRecordBatchStream> {
-        Ok(Box::pin(FilterExecStream {
-            schema: self.input.schema().clone(),
-            predicate: self.predicate.clone(),
-            input: self.input.execute(partition).await?,
-        }))
+        let predicate = self.predicate.clone();
+
+        let stream = self.input.execute(partition).await?;
+        let stream = stream.then(move |batch| {
+            let predicate = predicate.clone();
+            async move {
+                // Filtering batches is CPU-bounded and therefore justifies a dedicated thread pool
+                task::spawn_blocking(move || batch_filter(&batch?, &predicate))
+                    .await
+                    .unwrap()
+            }
+        });
+
+        Ok(Box::pin(SizedRecordBatchStream::new(stream, self.schema())))
```

Comment:
This is the core idea: wrap the input stream with a blocking thread that runs `batch_filter` (the offending blocking code).

Reply:
Does this code run multiple `batch_filter` for different batches at the same time with this code?

Reply:
It looks to me it currently does not?

Reply:
no, but this did not happen in master either: a stream does not ask for the next item until the current item is completed. We get no parallelism whatsoever from streams, only concurrency (because they yield). E.g. https://docs.rs/parallel-stream/2.1.2/parallel_stream/#differences-with-rayon

Reply:
I think buffering with a channel is the Rust-y  way to get more parallel execution without unlimited buffering

Reply:
Thanks @jorgecarleitao this seems to make a lot of sense.

@alamb Was your comment a response to @Dandandan or an objection to using `spawn_blocking`? I wasn't sure.

Reply:
My comment regarding channels was a response to @Dandandan  and @jorgecarleitao 's discussion of potentially getting more parallelism going. My comments regarding the threadpools was in response to this PR in general